### PR TITLE
plugin/cache: eliminate stale cache refresh and prefetch race

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. Why is this pull request needed and what does it do?

This addresses the redundant fetch and race issue raised in #4324 and #4364.
It adds a test that detects the race condition in the existing code, while passing with these changes.
It consolidates stale cache async refresh and the async prefetch to use the same prefetch code.
Doing this normalizes some different behavior between refresh and prefetch:
1. Do bit is now always set.  Previously it was set during cache refresh, but not prefetch.
2. RemoteIP is used as the remoteIP during cache refresh. Previously it was set to LocalIP during cache refresh.

### 2. Which issues (if any) are related?

#4324, #4364

### CODE REVIEW

**Comment by chrisohaver** [comment]
Another thing this changes:  It ticks the prefetch count metric up when fetching a stale cache cache item.  IMO, this is OK, since it is a kind of "prefetch" in itself.

**Comment by sschepens** [comment]
@chrisohaver i hadn't seen this PR and i pushed one myself #4368

I can see we took different approaches.

To make sure we don't hit #4271 we should probably move `newPrefetchResponseWriter` creation outside the goroutine.

File: plugin/cache/prefetch_test.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ func TestPrefetch(t *testing.T) {
 		qname         string
 		ttl           int
 		prefetch      int
+		staleUpTo     time.Duration
```
Comment (miekg) [inline plugin/cache/prefetch_test.go:21]: where is this used? I only see assignments to staleUpTo ?

Comment (chrisohaver) [reply plugin/cache/prefetch_test.go:21]: used here... https://github.com/coredns/coredns/pull/4367/files#diff-b6eb3644f59cc77177fa546593ae1b637cfb4699ba5b1664df21574eb874e2acR125

**Comment by chrisohaver** [comment]
duplicate of #4368

File: plugin/cache/prefetch_test.go (hunk 1 above)
Comment (miekg) [reply plugin/cache/prefetch_test.go:21]: I see. lgtm

**Review by miekg — APPROVED** [review]

**Pull request closed without merging**